pkg/agent/endpoints/sdsv3/handler.go

```diff
@@ -307,6 +337,171 @@ func (h *Handler) triggerReceivedHook() {
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to parse bundle's trust domain: %v", err)
+		}
+		bundles[td] = tdBundle
+	}
+
+	// Add all federated bundles
+	for td, bundle := range federatedBundles {
+		bundles[td] = bundle
+	}
+
+	return &spiffeBuilder{
+		bundles: bundles,
+	}, nil
+}
+
+func (b *spiffeBuilder) buildOne(resourceName, trustDomainID string) (*any.Any, error) {
+	td, err := spiffeid.TrustDomainFromString(trustDomainID)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to parse trustdomain: %v", err)
+	}
+	bundle, ok := b.bundles[td]
+	if !ok {
+		return nil, status.Errorf(codes.Internal, "no bundle found for trust domain: %q", trustDomainID)
+	}
+
+	caBytes := pemutil.EncodeCertificates(bundle.RootCAs())
+	typedConfig, err := anypb.New(&tls_v3.SPIFFECertValidatorConfig{
+		TrustDomains: []*tls_v3.SPIFFECertValidatorConfig_TrustDomain{
+			{
+				Name: td.String(),
+				TrustBundle: &core_v3.DataSource{
+					Specifier: &core_v3.DataSource_InlineBytes{
+						InlineBytes: caBytes,
+					},
+				},
+			},
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return anypb.New(&tls_v3.Secret{
+		Name: resourceName,
+		Type: &tls_v3.Secret_ValidationContext{
+			ValidationContext: &tls_v3.CertificateValidationContext{
+				CustomValidatorConfig: &core_v3.TypedExtensionConfig{
+					Name:        "envoy.tls.cert_validator.spiffe",
+					TypedConfig: typedConfig,
+				},
+			},
+		},
+	})
+}
+
+func (b *spiffeBuilder) buildAll(resourceName string) (*any.Any, error) {
+	configTrustDomains := []*tls_v3.SPIFFECertValidatorConfig_TrustDomain{}
+
+	// Create SPIFFE validator config
+	for td, bundle := range b.bundles {
+		// bundle := bundles[td]
+		caBytes := pemutil.EncodeCertificates(bundle.RootCAs())
+		configTrustDomains = append(configTrustDomains, &tls_v3.SPIFFECertValidatorConfig_TrustDomain{
+			Name: td.String(),
+			TrustBundle: &core_v3.DataSource{
+				Specifier: &core_v3.DataSource_InlineBytes{
+					InlineBytes: caBytes,
+				},
+			},
+		})
+	}
+
+	// // Order by trustdomain name to return in consistent order
+	sort.Slice(configTrustDomains, func(i, j int) bool {
+		return configTrustDomains[i].Name < configTrustDomains[j].Name
+	})
+
+	typedConfig, err := anypb.New(&tls_v3.SPIFFECertValidatorConfig{
+		TrustDomains: configTrustDomains,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return anypb.New(&tls_v3.Secret{
+		Name: resourceName,
+		Type: &tls_v3.Secret_ValidationContext{
+			ValidationContext: &tls_v3.CertificateValidationContext{
+				CustomValidatorConfig: &core_v3.TypedExtensionConfig{
+					Name:        "envoy.tls.cert_validator.spiffe",
+					TypedConfig: typedConfig,
+				},
+			},
+		},
+	})
+}
+
+func supportsSPIFFEAuthExtension(req *discovery_v3.DiscoveryRequest) bool {
+	if buildVersion := req.Node.GetUserAgentBuildVersion(); buildVersion != nil {
+		version := buildVersion.Version
+		return version.MajorNumber >= 1 && version.MinorNumber > 17
```

Comment:
This check is potentially incorrect for Envoy 2.0 - 2.17 (which may or may not support the extension, but we can assume they do until we know better).
```suggestion
		return (version.MajorNumber == 1 && version.MinorNumber > 17) || version.MajorNumber > 1
```